core/state/state_object.go

```diff
@@ -233,7 +240,9 @@ func (c *StateObject) AddBalance(amount *big.Int) {
 }
 
 func (c *StateObject) SubBalance(amount *big.Int) {
-	if amount.Cmp(common.Big0) == 0 {
+	// We must check emptiness for the objecst such that the account
+	// clearing (0,0,0 objects) can take effect.
+	if amount.Cmp(common.Big0) == 0 && c.empty() {
```

Comment:
@I was looking at my code wondering why it said `c.empty()` here and `!c.empty()` on my computer, turns out this was also added for `SubBalance` which shouldn't be here (although it doesn't really quite matter). I'm removing the `empty` check from `SubBalance`.